backend/wmg/data/query.py

```diff
@@ -0,0 +1,98 @@
+from typing import List
+
+import tiledb
+from pandas import DataFrame
+from pydantic import BaseModel, Field
+from tiledb import Array
+
+from backend.wmg.data.ontology_labels import gene_term_label, ontology_term_label
+
+ALL_DIM_VALUES = slice(None)
+EMPTY_DIM_VALUES = ""
+
+
+class WmgQueryCriteria(BaseModel):
+    gene_ontology_term_ids: List[str] = Field(default=[], unique_items=True, min_items=0)
+    organism_ontology_term_id: str  # required!
+    tissue_ontology_term_ids: List[str] = Field(unique_items=True, min_items=1)  # required!
+    dataset_ids: List[str] = Field(default=[], unique_items=True, min_items=0)
+    # excluded per product requirements, but keeping in, commented-out, to reduce future head-scratching
+    # assay_ontology_term_ids: List[str] = Field(default=[], unique_items=True, min_items=0)
+    development_stage_ontology_term_ids: List[str] = Field(default=[], unique_items=True, min_items=0)
+    disease_ontology_term_ids: List[str] = Field(default=[], unique_items=True, min_items=0)
+    ethnicity_ontology_term_ids: List[str] = Field(default=[], unique_items=True, min_items=0)
+    sex_ontology_term_ids: List[str] = Field(default=[], unique_items=True, min_items=0)
+
+
+indexed_dims = {"gene_ontology_term_ids", "organism_ontology_term_id", "tissue_ontology_term_ids"}
+
+
+class WmgQuery:
+    def __init__(self, cube: Array) -> None:
+        super().__init__()
+        self._cube = cube
+
+    def execute(self, criteria: WmgQueryCriteria) -> DataFrame:
+        # As TileDB API does not yet support logical OR'ing of attribute values in query conditions, the best we can do
+        # for a single TileDB query is to have TileDB perform the filtering for only the attributes that have
+        # a single criterion value specified by the user. We can then perform the filtering client-side for the
+        # multi-valued attributes, albeit at the cost of retrieving extra data from the TileDB engine. Note, however,
+        # that this extra data is always being retrieved by TileDB from the filesystem (and so across the network), so
+        # the performance impact should not be as bad as it may at first appear.
+
+        # Filter single-valued attribute criterion using TileDB filtering, using QueryCondition
+        # TODO: Benchmark whether this is faster than just doing all the attribute filtering in Pandas
+        single_valued_attrs = {
```

Comment:
This code might be more clear if you pulled the single valued attr query building and the multi valued attr query building into separate functions